qmail-local.c

```diff
@@ -92,21 +108,37 @@ char *dir;
  pid = getpid();
  host[0] = 0;
  gethostname(host,sizeof(host));
+
```

Comment:
one could do an initial allocation for hostname here, so it has enough room to hold "host" first, which should be enough for all sane cases anyway. Or maybe just do a scan through host if it has any of the special chars and just copy over the whole string otherwise (i.e. always).

Reply:
POSIX gethostname does not null-terminate the buffer if it was too long, so I recommend:
```
        host[sizeof host - 1] = 0;
```

Reply:
The best solution here is for users to adopt OpenBSD's gethostname(3): "The returned name is always NUL terminated." I've updated this to use a stralloc and then store the hostname in the sa.s array and write '\0' in the last byte. For hostnames with 256 or more characters, this will truncate it to the first 255.

Reply:
This still misses to preallocate hostname as shown above.

One could also think of not preallocating and simply not touching hostname at all until we actually found a position where it needs to be recoded, and only then allocate it and copy the whole prefix at once. This would however make the code slightly more complicated.